File: capitolo_1/octave/Esercizio1_6.m
function Esercizio1_6()
  close all
  clear all
  clc
  % Si applichi il controllore a relè con isteresi per il quale e = hbar - h(t).
  % Si faccia riferimento al serbatoio dell'esempio 1.13.
  % Si studi l'evoluzione h(t) per h(0)=h0 generico al variare di Ehat, qhat.

  %
  % Dati 1.17: Processo di riempimento ad isteresi con portata costante qhat:
  %   A * hdot(t) = qhat , e>Ehat
  %               = -qhat , e<Ehat
  % Nel transitorio iniziale si fa riferimento all'andamento della variazione
  % di livello
  %
  A = 0.5; % Area iniziale
  qbar = 2; % Portata discreta
  Ebar = 0.5; % Margine d'errore per isteresi
  hbar = 3; % Target di livello
  h0 = 0;   % Livello iniziale
  qprev = qbar; %Variabile di appoggio per gestione isteresi portata\

  dt = 0.001; %Definizione del tempo
  ti = 0;     %Tempo inizio
  tf = 3;     %Tempo fine
  t = ti:dt:tf;  %Vettore tempi
  q = zeros(size(t)); %Vettore portata
  h = ones(size(t)) * h0; %Vettore livello

  for i = 1 : length(t)-1
    e = hbar - h(i);
    if (e >= Ebar)
      q(i) = qbar;
    elseif (e <= -Ebar)
      q(i) = -qbar;
    else
      q(i) = qprev;
    endif
    qprev = q(i);
    hdot = q(i)/A;
    h(i+1) = h(i) + hdot * dt;
  endfor

  figure(1)
  plot(t,h, [ti tf], [hbar hbar], 'k');
  ylim([0 4]);
  grid;
  xlabel('Tempo (t)')
  ylabel('Livello h(t)')
  title('Evoluzione del livello h(t) per C.I. nulle')

  figure(2)
  plot(t,q, [ti tf], [0 0], 'k');
  ylim([-3 3]);
  grid;
  xlabel('Tempo (t)')
  ylabel('Portata q(t)')
  title('Evoluzione della portata q(t) per C.I. nulle')

endfunction

File: capitolo_1/cpp/Esercizio1_6.cpp
#include <iostream>
#include <vector>
#include <string>
#include "matplotlibcpp.h"

namespace plt = matplotlibcpp;

// --------------------------
// MAIN
// --------------------------
int main() {
    // Dati
    double A = 0.5, hbar = 3, qbar = 2, Ebar = 0.5, h0 = 0, qprev = qbar;
    double dt = 0.001;
    double t0 = 0, tf = 3;
    // Vettori per i risultati
    int N = static_cast<int>((tf - t0) / dt);
    std::vector<double> time(N + 1);
    for (int i = 0; i <= N; ++i) time[i] = t0 + i * dt;
    std::vector<double> h(N + 1);
    std::vector<double> q(N + 1);
    h[0] = h0; //Impongo la condizione iniziale

    // -----------------------------------------
    // Figura 1.6 - Evoluzione temporale del livello dell'acqua nel serbatoio
    // -----------------------------------------
    for (int i = 0; i < N; ++i) {
        double errore = hbar - h[i];
        if (errore >= Ebar)
            q[i] = qbar;
        else if (errore <= -Ebar)
            q[i] = -qbar;
        else
            q[i] = qprev;
        qprev = q[i];
        // Calcolo della derivata di h(t)
        double hdot = q[i] / A;
        // Calcolo del nuovo valore di h(t+dt)
        h[i + 1] = h[i] + hdot * dt;
    }

    
    // Plot Figura 1.6 - Livello dell'acqua nel serbatoio
    plt::figure(1);
    plt::plot(time, h, {{"label", "h(t)"}});
    plt::xlabel("Tempo (t)");
    plt::ylabel("Livello dell'acqua h");
    plt::title("Evoluzione temporale del livello dell'acqua nel serbatoio");
    plt::legend();
    plt::grid(true);

    // Plot Figura 1.6b - Portata
    plt::figure(2);
    plt::plot(time, q, {{"label", "q(t)"}});
    plt::xlabel("Tempo (t)");
    plt::ylabel("Portata q");
    plt::title("Evoluzione temporale della portata nel serbatoio");
    plt::legend();
    plt::grid(true);

    plt::show();

    return 0;
}
